### PR TITLE
[Feature][API][backend] Query all upstream and downstream dependencies of the task

### PR DESCRIPTION
**Pull request by IamMujuziMoses**
## What is the purpose of the pull request

*Fixes issue #3797*
- [x] Upstream Dependency
- [x] Downstream Dependency
- [ ] Upstream and Downstream dependency

## Brief change log

*ProcessDefinitionController.java*
*ProcessDefinitionService.java*
*ProcessDefinitionServiceImpl.java*
  - *Added method `queryUpstreamTaskDependencies`*
  - *Added method `queryDownstreamTaskDependencies`*

*ProcessDefinitionControllerTest.java*
*ProcessDefinitionServiceTest.java*
 - *Added method `testQueryUpstreamTaskDependencies`*
 - *Added method `testQueryDownstreamTaskDependencies`*

*DependentParameters.java*
*DependentItem.java*
*DependentTaskModel.java*
 - *Added `toString()` method*

*Status.java*
 - *Added `QUERY_TASK_DEPENDENCIES_BY_TASK_NAME_ERROR`*

### CODE REVIEW

**Comment by CalvinKirs** [comment]
hi, deeply thanks for your contribution, but you need to resolve the conflicting files, otherwise you cannot merge.

**Comment by IamMujuziMoses** [comment]
@CalvinKirs @lgcareer @dailidong This is the initial PR to solving issue #3797, please review. Is the feature design suggestion okay?! or should I change it?!

File: dolphinscheduler-api/src/test/java/org/apache/dolphinscheduler/api/service/ProcessDefinitionServiceTest.java [hunk 1]
```diff
@@ -1329,4 +1329,9 @@ public void testImportProcessSchedule() {
         Assert.assertEquals(0, processDefinitionService.importProcessSchedule(loginUser, projectName, processMeta, processDefinitionName, processDefinitionId));
     }
 
+    @Test
+    public void  testQueryTaskDependenciesByTaskName() {
```
Comment (chengshiwen) [inline dolphinscheduler-api/src/test/java/org/apache/dolphinscheduler/api/service/ProcessDefinitionServiceTest.java]: You should provide UT for `testQueryTaskDependenciesByTaskName`

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/ProcessDefinitionServiceImpl.java [hunk 2]
```diff
@@ -1836,5 +1836,10 @@ private void checkBatchOperateResult(String srcProjectName, String targetProject
         }
     }
 
+    public Map<String, Object> queryTaskDependenciesByTaskName(User loginUser, String projectName, Integer processId, String taskName) {
+
+        return null;
```
Comment (chengshiwen) [inline dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/ProcessDefinitionServiceImpl.java]: The work seems to be unfinished

Comment (CalvinKirs) [reply dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/ProcessDefinitionServiceImpl.java:1841]: well,can you finish it?

**Comment by lgcareer** [comment]
@IamMujuziMoses So sorry for response late.Hi,I saw here are several issue need to resolve.You need finished the implement code in the ProcessDefininitionImpl.

**Comment by IamMujuziMoses** [comment]
@lgcareer @CalvinKirs @chengshiwen Thanks for the review, am going to complete it and send  a commit today

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/ProcessDefinitionServiceImpl.java [hunk 3]
```diff
@@ -18,6 +18,7 @@
 package org.apache.dolphinscheduler.api.service.impl;
 
 import static org.apache.dolphinscheduler.common.Constants.CMD_PARAM_SUB_PROCESS_DEFINE_ID;
+import static org.apache.dolphinscheduler.dao.upgrade.UpgradeDao.getDataSource;
```
Comment (IamMujuziMoses) [inline dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/ProcessDefinitionServiceImpl.java]: This is not needed

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/ProcessDefinitionServiceImpl.java [hunk 4]
```diff
@@ -73,6 +74,7 @@
 import org.apache.dolphinscheduler.service.permission.PermissionCheck;
 import org.apache.dolphinscheduler.service.process.ProcessService;
 
+import javax.sql.DataSource;
```
Comment (IamMujuziMoses) [inline dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/ProcessDefinitionServiceImpl.java]: this too is not neeeded

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/ProcessDefinitionServiceImpl.java [hunk 5]
```diff
@@ -122,6 +124,8 @@
 
     private static final String TASKS = "tasks";
 
+    protected static final DataSource dataSource = getDataSource();
```
Comment (IamMujuziMoses) [inline dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/ProcessDefinitionServiceImpl.java]: This too is not needed

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/ProcessDefinitionServiceImpl.java [hunk 6]
```diff
@@ -1836,5 +1840,40 @@ private void checkBatchOperateResult(String srcProjectName, String targetProject
...
+        }
+
+        List<TaskNode> taskNodeList = (processData.getTasks() == null) ? new ArrayList<>() : processData.getTasks();
+
+        for (TaskNode taskNode : taskNodeList) {
+            if (taskNode.getName().equals(taskName)) {
+                depList = taskNode.getDepList();
+            }
+        }
+        result.put(Constants.DATA_LIST, depList);
```
Comment (lgcareer) [inline dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/ProcessDefinitionServiceImpl.java:1925]: Hi,you should judge the task node type,if it's type is `DEPENDENT`,you should parse the field `dependence` to get the upstream dependences,and they are some process definitions or some tasks.

**Comment by IamMujuziMoses** [comment]
Hi @lgcareer @dailidong sonar is reporting some code smell and vulnerability but for the code smell, that is the needed return type and for the vulnerability, we need to use the user-controlled data, so that is a mistake in the sonar results

**Comment by CalvinKirs** [comment]
> Hi @lgcareer @dailidong sonar is reporting some code smell and vulnerability but for the code smell, that is the needed return type and for the vulnerability, we need to log the user-controlled data as protocol, so that is a mistake in the sonar results

Very good, I will review for you later

**Comment by dailidong** [comment]
> Hi @lgcareer @dailidong sonar is reporting some code smell and vulnerability but for the code smell, that is the needed return type and for the vulnerability, we need to log the user-controlled data as protocol, so that is a mistake in the sonar results

yes, you are right, we will ignore the vulnerability reported by sonarcloud

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/ProcessDefinitionServiceImpl.java (hunk 6 above)
Comment (lgcareer) [reply dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/ProcessDefinitionServiceImpl.java:1925]: Hi,you should parse the field dependence.

**Comment by lgcareer** [comment]
@IamMujuziMoses  Great Job,The design seems no problem,I think the downstream dependency is OK,I will check it carefully again, If you want to expand the feature,you can think how to get the upstream dependency.

**Comment by dailidong** [comment]
hi, @IamMujuziMoses , please solve the conflict files locally
image

**Comment by IamMujuziMoses** [comment]
@CalvinKirs @chengshiwen @Segun-Ogundipe  need some help here, with the Unit Test/Build, what might be causing it to run endlessly?!

**Comment by davidzollo** [comment]
> @CalvinKirs @chengshiwen @Segun-Ogundipe need some help here, with the Unit Test/Build, what might be causing it to run endlessly?!

I will try restart the UT test

**Comment by IamMujuziMoses** [comment]
Let me close this PR and open another one #4778

**Pull request closed without merging**